feat: extract optimism genesis info

Vid201 opened the pull request:
## Motivation

Adds functionalities to extract Optimism genesis data from genesis.

`extract_from` returns `Option` since the extraction can be unsuccessful - negative case should be handled outside.

mattsse commented:
yeah, having the sanity test would be great

mattsse commented on crates/rpc-types/src/genesis.rs:
```diff
@@ -1,5 +1,7 @@
 //! OP types for genesis data.
 
+use alloy_genesis::Genesis;
```
could we import Otherfields from alloy serde instead of depending on genesis?

mattsse commented on crates/rpc-types/src/genesis.rs:
```diff
@@ -27,3 +29,22 @@ pub struct OptimismBaseFeeInfo {
+
+impl OptimismGenesisInfo {
+    /// Extract the Optimism genesis info from the genesis.
+    pub fn extract_from(genesis: &Genesis) -> Option<Self> {
```
could we add another helper type that has both genesisInfo and feeinfo as fields and then impl fn extract_from(others: &Otherfields)?

and also `Tryfrom<Otherfields>` that simply calls extract?

Vid201 commented:
Addressed the comments

mattsse commented on crates/rpc-types/src/genesis.rs:
```diff
@@ -27,3 +79,172 @@ pub struct OptimismBaseFeeInfo {
+    type Error = serde_json::Error;
+
+    fn try_from(others: OtherFields) -> Result<Self, Self::Error> {
+        if let Some(Ok(optimism_base_fee_info)) =
+            others.get_deserialized::<OptimismBaseFeeInfo>("optimism")
+        {
+            Ok(optimism_base_fee_info)
```
this code is duplicated, this should reuse the extract fn

Vid201 replied:
But if we reuse the extract fn we lose the error information (since extract returns option) - thus cannot use `type Error = serde_json::Error`. or should we define custom error enum?

mattsse replied:
I see, then you can flip it the other way around

mattsse commented on crates/rpc-types/src/genesis.rs:
```diff
@@ -27,3 +79,172 @@ pub struct OptimismBaseFeeInfo {
 }
+
+impl OptimismBaseFeeInfo {
+    /// Extract the Optimism base fee info from `OtherFields` object.
```
this should document this will check for an `optimism` object
and is intended to be parsed from the genesis object

mattsse commented on crates/rpc-types/src/genesis.rs:
```diff
@@ -16,6 +53,21 @@ pub struct OptimismGenesisInfo {
+    type Error = serde_json::Error;
+
+    fn try_from(others: OtherFields) -> Result<Self, Self::Error> {
+        others.deserialize_as()
```
should delegate to Self::extract_from

mattsse commented on crates/rpc-types/src/genesis.rs:
```diff
@@ -1,5 +1,42 @@
+    type Error = serde_json::Error;
+
+    fn try_from(others: OtherFields) -> Result<Self, Self::Error> {
+        let genesis_info = OptimismGenesisInfo::try_from(others.clone())?;
+        let base_fee_info = OptimismBaseFeeInfo::try_from(others)?;
+
+        Ok(OptimismChainInfo {
+            genesis_info: Some(genesis_info),
+            base_fee_info: Some(base_fee_info),
+        })
```
reuse existing impl

mattsse commented on crates/rpc-types/src/genesis.rs:
```diff
@@ -1,5 +1,42 @@
+use alloy_serde::OtherFields;
+use serde::de::Error;
+
+/// Info for Optimism chain.
```
this should document that this is a container type for all OP specific fields in a genesis file.

mattsse commented on crates/rpc-types/src/genesis.rs:
```diff
@@ -1,5 +1,42 @@
+}
+
+impl OptimismChainInfo {
+    /// Extract the Optimism chain info from `OtherFields` object.
```
needs additional context that this is expected to be other, optimism specific fields of `genesis.config`

mattsse reviewed (changes requested):
great, a few more nits.

the tryinto are more or less just for convenience but should reuse the existing functions so we don't have the code duplicated.

Vid201 commented:
> great, a few more nits.
> 
> the tryinto are more or less just for convenience but should reuse the existing functions so we don't have the code duplicated.

@mattsse Addressed the comments

mattsse reviewed (approved):
nice
